thoth/package_extract/image.py

```diff
@@ -463,6 +463,32 @@ def construct_rootfs(dir_path: str, rootfs_path: str) -> list:
     return layers
 
 
+def _get_python_version(path: str) -> dict:
+    python_version = {
+        "python_interpreters": [],
+        "default_interpreters": {}
+    }
+
+    cmd = "ls -la {!r}/usr/bin/python*".format(path)
```

Comment:
It would be better to check all Python binaries in the path and traverse symlinks in Python code (you can use [globbing](https://docs.python.org/3/library/glob.html) and [os.path.islink](https://docs.python.org/3/library/os.path.html#os.path.islink) to check if the file is a symlink). BTW you will need to simulate symlink traversals, symlinks will refer to other symlinks often - e.g. on my fedora:30:

```
$ ls -la /usr/bin/python
lrwxrwxrwx. 1 root root 7 Apr 30 17:55 /usr/bin/python -> python2
$ ls -la /usr/bin/python2
lrwxrwxrwx. 1 root root 9 Apr 30 17:55 /usr/bin/python2 -> python2.7
$ ls -la /usr/bin/python2.7
-rwxr-xr-x. 1 root root 18720 Apr 30 17:55 /usr/bin/python2.7
```

You can reuse [MAXSYMLINKS constant from the Linux kernel](https://github.com/torvalds/linux/blob/54ecb8f7028c5eb3d740bb82b0f1d90f2df63c5c/include/linux/namei.h#L12) to stop symlink traversal in order to eliminate a symlinks-loop.

Reply:
@fridex I am curious what output in the JSON format we want see for this particular case for example:
`
 bdauletb@localhost  ~/D/storages   sync-duration  ls -la /usr/bin/python*
lrwxrwxrwx. 1 root root     7 Apr 30 11:55 /usr/bin/python -> python2
lrwxrwxrwx. 1 root root     9 Apr 30 11:55 /usr/bin/python2 -> python2.7
-rwxr-xr-x. 1 root root 18720 Apr 30 11:55 /usr/bin/python2.7
lrwxrwxrwx. 1 root root     9 Mar 27  2019 /usr/bin/python3 -> python3.7
-rwxr-xr-x. 2 root root 16680 Feb 19  2019 /usr/bin/python3.6
lrwxrwxrwx. 1 root root    17 Feb 19  2019 /usr/bin/python3.6-config -> python3.6m-config
-rwxr-xr-x. 2 root root 16680 Feb 19  2019 /usr/bin/python3.6m
-rwxr-xr-x. 1 root root   177 Feb 19  2019 /usr/bin/python3.6m-config
-rwxr-xr-x. 1 root root  3628 Feb 19  2019 /usr/bin/python3.6m-x86_64-config
-rwxr-xr-x. 2 root root 16680 Mar 27  2019 /usr/bin/python3.7
lrwxrwxrwx. 1 root root    17 Mar 27  2019 /usr/bin/python3.7-config -> python3.7m-config
-rwxr-xr-x. 2 root root 16680 Mar 27  2019 /usr/bin/python3.7m
-rwxr-xr-x. 1 root root   177 Mar 27  2019 /usr/bin/python3.7m-config
-rwxr-xr-x. 1 root root  3636 Mar 27  2019 /usr/bin/python3.7m-x86_64-config
lrwxrwxrwx. 1 root root    16 Mar 27  2019 /usr/bin/python3-config -> python3.7-config
-rwxr-xr-x. 1 root root   387 Feb  2  2019 /usr/bin/python3-coverage
-rwxr-xr-x. 1 root root   396 Feb  2  2019 /usr/bin/python3-mako-render
`

Reply:
@fridex Just trying to understand why we should use `glob`. It is good for me to use `glob` to find all locations at `/usr/bin/python*`. However, it looks like it does not change anything it terms what `ls -la` finds.
Second, is there really need to do symlink traversal? From what I see it seems to me we can easily get that info and symlinks are sort of finite. The longest one I see like `python -> python2 -> python2.7` and there is no further, the same applies for python3.
Third, I'm sort of confused what needs to be returned `python version` or `python interpreter locations`?

Reply:
> @fridex Just trying to understand why we should use `glob`. It is good for me to use `glob` to find all locations at `/usr/bin/python*`. However, it looks like it does not change anything it terms what `ls -la` finds.

The overall result is pretty much same, except for the overhead needed to run `ls` - spawning a new sub-process, executing it and gathering+parsing all the data instead of using directly routines which Python provides which will directly talk to the kernel to obtain this info.

> Second, is there really need to do symlink traversal? From what I see it seems to me we can easily get that info and symlinks are sort of finite. The longest one I see like `python -> python2 -> python2.7` and there is no further, the same applies for python3.

This might be on your system, but we don't have control of what is being analyzed - and it can differ from container image (in most cases there will be top 2 links as you write, but just to be sure...).

> Third, I'm sort of confused what needs to be returned `python version` or `python interpreter locations`?

Returning locations - actual Python versions can be obtained from rpm/deb package discovery we already do.

We can start with something like:
```json
{
  "python_interpreters": [
    {"path": "/usr/bin/python"},
    {"path": "/usr/bin/python2"},
    {"path": "/usr/bin/python2.7"},
    {"path": "/usr/bin/python3"},
    {"path": "/usr/bin/python3.6"},
    {"path": "/usr/bin/python3.6-config"},
    {"path": "/usr/bin/python3.6m"},
    {"path": "/usr/bin/python3.6m-config"},
    {"path": "/usr/bin/python3.6m-x86_64-config"},
    {"path": "/usr/bin/python3.7"},
    {"path": "/usr/bin/python3.7-config"},
    {"path": "/usr/bin/python3.7m"},
    {"path": "/usr/bin/python3.7m-config"},
    {"path": "/usr/bin/python3.7m-x86_64-config"},
    {"path": "/usr/bin/python3-config"},
    {"path": "/usr/bin/python3-coverage"},
    {"path": "/usr/bin/python3-mako-render"},
  ],
}
```

In the next iteration, you can add information about symlinks. The `default_interpreters` as discussed above can be later obtained explictly if we design `python_interpreters` in the right way (so no need to provide it explicitly),